### PR TITLE
sql: don't try to add index in ALTER...ADD FK

### PR DESCRIPTION
**Pull request by dt**
ALTER...ADD FOREIGN KEY reuses the index resolution and modification logic used in CREATE TABLE...FOREIGN KEY.

However, as of 668f3c74, that logic included *adding* the required source index to the table if it didn't exist. This is fine during CREATE TABLE, but is *not* fine for an existing table, when index additions need to go though the schema change process to be staged, backfilled, etc correctly.

### CODE REVIEW

**Comment by dt** [comment]
This appears to break the ActiveRecord example app's migration.

Currently I think we should merge it anyway -- the behavior that was making the ActiveRecord example appear to work was, and is, wrong, and we should stop doing it ASAP.

That said, why the active record migration is broken is another story: it runs all the DDLs in its migration in one commit, and even though the ruby (if you can call it that) version of orders says `t.belongs_to :customer, index: true, null: false`, that `index: true` gets translated into a separate, followup `CREATE INDEX ON ...`, and then, due to our ddl-in-txn complications, it looks like that index doesn't actually appear in the table descriptor when the next `ALTER TABLE ADD ... FOREIGN KEY` gets run, causing it to (want to) spuriously add the same index again.
Indeed, in the schema, you'll see all the fk indexes are duplicated -- one created by rails, one by cockroach when it can't see the rails one.

**Comment by dt** [comment]
I filed #16182 for the discussion of the the `add index` in same txn not showing up, which, if this merges as it, will be what's breaking active record.

File: pkg/sql/create.go
Comment (vivekmenezes) [inline pkg/sql/create.go]: I'm confused as to why this error is related to the mode. When does this mode get set, and why is unsetting it okay?

Comment (dt) [inline pkg/sql/create.go]: the mode is passed as an argument to resolveFK, to `validated` when it is called in during `CREATE` and `unvalidated` when called in `ALTER`

**Comment by vivekmenezes** [comment]
LGTM

**Comment by dt** [comment]
I think we should probably cherry-pick this one -- the correctness issue leaving it as-is is pretty scary.

File: pkg/sql/create.go
Comment (dt) [inline pkg/sql/create.go]: Hm. I'm not sure I follow.. I assume a user wouldn't *expect* to see an error at all, if they didn't realize they need an index on the fK columns?

Comment (jordanlewis) [inline pkg/sql/create.go]: Sorry, that wasn't very clear. What I meant to say is that this error message seems too low-level. If I were a user who is trying to add an FK column that doesn't have a corresponding index yet, seeing this message would not help me figure out how to resolve my problem.

How about something along the lines of `cannot add foreign key columns %s without a pre-existing index`?

Comment (dt) [inline pkg/sql/create.go]: sure. one thing to note is that a pre-existing index need only have the columns they're trying to use as a *prefix* (since that is good enough for retrieval), unlike the referenced index, which we, and postgres, require be unique, and thus must match the constrained columns exactly. 

Do we feel like trying to clarify this subtly in a error message? or should we just let the docs take over here if someone really wants to get into he details?

Comment (jordanlewis) [inline pkg/sql/create.go]: I'm fine with not clarifying the subtlety perfectly but I think it's important that the error message make it obvious what exactly is going wrong and how to resolve the problem. I'll leave it up to you to decide whether this is sufficient - I just expect that this error message will not be particularly enlightening for users that encounter it.

Comment (dt) [inline pkg/sql/create.go]: reworked, RFAL.

**Review by jordanlewis — APPROVED** [review]
thanks!

**Comment by dt** [comment]
Okay, just blocked on cockroachdb/examples-orms#27 now I think.

**Comment by mberhault** [comment]
FYI: this has not been cherry-picked for `1.0.2` yet.

**Pull request merged**
